### PR TITLE
[LPT] Q/DQ support

### PR DESCRIPTION
**Pull request by eshoguli**
validation:
* developer-services/job/accuracy/1087/
* developer-services/job/performance/562/ (developer-services/job/performance/552/ - on previous models)

validation:
* developer-services/job/accuracy/1015/
* developer-services/job/performance/535/ (developer-services/job/performance/543/)

validation:
* developer-services/job/accuracy/969/ (AAQ, not limited by time)
* developer-services/job/accuracy/968/ (AAQ, aborted)
* developer-services/job/performance/509/

validation:
* developer-services/job/accuracy/961/ (default)
* developer-services/job/accuracy/960/ (AAQ)

validation:
* developer-services/job/accuracy/932/ (default) (reference: developer-services/job/accuracy/950/)
* developer-services/job/accuracy/934/ (AAQ) (reference: developer-services/job/accuracy/951/)
* developer-services/job/performance/499/

validation:
* developer-services/job/accuracy/921
* developer-services/job/performance/495 (reference: developer-services/job/performance/496/)

validation:
* AVX 512 VNNI: developer-services/job/accuracy/915/
* developer-services/job/performance/493/ (developer-services/job/performance/494/)

validation:
* AVX512: developer-services/job/accuracy/871/ (reference in FP32: developer-services/job/accuracy/872/), AVX512 VNNI: general pool: developer-services/job/accuracy/903, custom pool: developer-services/job/accuracy/906/ (reference in FP32: developer-services/job/accuracy/905/, custom pool: developer-services/job/accuracy/907/)
* developer-services/job/performance/482/ 

validation:
* accuracy: developer-services/job/accuracy/854/ (reference (previous models + INT8): developer-services/job/accuracy/855, reference (new models + FP32): developer-services/job/accuracy/856/). Master on previous models in FP32: default: developer-services/job/accuracy/863/ (AAQ: developer-services/job/accuracy/857/)
* performance: developer-services/job/performance/473/ (reference: developer-services/job/performance/475/)

### CODE REVIEW

File: inference-engine/src/cldnn_engine/cldnn_engine.cpp [hunk 1]
```diff
@@ -208,9 +222,21 @@ InferenceEngine::ICNNNetwork::Ptr clDNNEngine::CloneAndTransformNetwork(const In
                         }
                         return true;
                     });
+
+            if (enableInt8) {
+                pass_config->set_callback<ngraph::pass::ConvertQuantizeDequantize>([](const_node_ptr &node) -> bool {
+                    const std::shared_ptr<ngraph::Node> multiply = const_cast<ngraph::Node*>(node.get())->shared_from_this();
```
Comment (vladimir-paramuzov) [inline inference-engine/src/cldnn_engine/cldnn_engine.cpp:228]:
```
auto multiply = std::const_pointer_cast<ngraph::Node>(node);
```

And why do you even need to have it non-constant in transformation callback?

**Review by vladimir-paramuzov — CHANGES_REQUESTED** [review]

File: inference-engine/src/cldnn_engine/cldnn_engine.cpp [hunk 2]
```diff
@@ -144,6 +149,8 @@ InferenceEngine::ICNNNetwork::Ptr clDNNEngine::CloneAndTransformNetwork(const In
                    std::dynamic_pointer_cast<const ::ngraph::opset5::LogSoftmax>(node);
         };
         auto nGraphFunc = clonedNetwork->getFunction();
+        // ngraph::pass::VisualizeTree("c:\\Projects\\temp\\gpu.original").run_on_function(nGraphFunc);
```
[the remark this answers is not in the document]
Comment (eshoguli) [reply inference-engine/src/cldnn_engine/cldnn_engine.cpp]: done, thanks

File: ngraph/core/include/ngraph/variant.hpp [hunk 3]
```diff
@@ -34,6 +34,7 @@ namespace ngraph
 
         virtual std::shared_ptr<ngraph::Variant> init(const std::shared_ptr<ngraph::Node>& node);
         virtual std::shared_ptr<ngraph::Variant> merge(const ngraph::NodeVector& nodes);
+        virtual std::string get_value() const;
```
Comment (ilyachur) [inline ngraph/core/include/ngraph/variant.hpp:37]: Why did you add this method?!

Comment (eshoguli) [reply ngraph/core/include/ngraph/variant.hpp]: debug info was removed, thanks

File: inference-engine/src/cldnn_engine/cldnn_engine.cpp (hunk 1 above)
Comment (eshoguli) [reply inference-engine/src/cldnn_engine/cldnn_engine.cpp]: Done, fixed:  was removed for `areQuantizeAndDequantizeSupportedForSubtract`, but was moved into `areQuantizeAndDequantizeSupportedForMultiply`. 
The type cast reason in `areQuantizeAndDequantizeSupportedForMultiply` method is nGraph LPT utils (similar as nGraph core utils: core\include\ngraph\graph_util.hpp) interfaces: `std::shared_ptr<ngraph::Node>` is used as input and output arguments.

File: inference-engine/src/transformations/include/transformations/low_precision/disable_convert_constant_folding_on_const_path.hpp [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+#include <memory>
+#include <utility>
+
+#include <transformations_visibility.hpp>
+#include <ngraph/pass/graph_rewrite.hpp>
+
+namespace ngraph {
+namespace pass {
+
+class TRANSFORMATIONS_API DisableConvertConcstantFoldingOnConstPath;
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/include/transformations/low_precision/disable_convert_constant_folding_on_const_path.hpp:16]: Why you put this transformation to transformations_library?

**Review by GlebKazantaev — CHANGES_REQUESTED** [review]

File: inference-engine/tests/functional/plugin/shared/src/low_precision_transformations/fake_quantize_with_dq_not_optimal_transformation.cpp [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        testValues.dequantizationOnWeights,
+        testValues.dequantizationAfter);
+}
+
+TEST_P(FakeQuantizeWithNotOptimalTransformation, CompareWithRefImpl) {
+    Run();
+
+    const auto actualType = getRuntimePrecision("output_original");
+    EXPECT_EQ(actualType, "U8");
+};
```
Comment (v-Golubev) [inline inference-engine/tests/functional/plugin/shared/src/low_precision_transformations/fake_quantize_with_dq_not_optimal_transformation.cpp:60]: Could you please override Run() to call kernel verification?

File: inference-engine/tests/functional/plugin/gpu/shared_tests_instances/low_precision_transformations/transpose_after_matmul_transformation.cpp [hunk 6]
```diff
@@ -23,7 +23,7 @@ const std::vector<bool> perTensorValues = { true, false };
 
 const std::vector<bool> transposeChannelDimValues = { true, false };
 
-INSTANTIATE_TEST_CASE_P(smoke_LPT, TransposeAfterMatMulTransformation,
+INSTANTIATE_TEST_CASE_P(DISABLED_smoke_LPT, TransposeAfterMatMulTransformation,
```
Comment (v-Golubev) [inline inference-engine/tests/functional/plugin/gpu/shared_tests_instances/low_precision_transformations/transpose_after_matmul_transformation.cpp:26]: We really need to disable these tests?

File: inference-engine/tests/functional/plugin/gpu/shared_tests_instances/low_precision_transformations/fake_quantize_with_dq_not_optimal_transformation.cpp [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+            { ngraph::element::f32, false },
+            { {127.f}, ngraph::element::f32, {}, false, 1ul, ngraph::element::i8, true },
+            { {0.3f}, ngraph::element::f32, {}, false }
+        },
+        {}
+    }
+};
+
+// TODO: add something to avoid cleanup and enable
+INSTANTIATE_TEST_CASE_P(DISABLED_smoke_LPT, FakeQuantizeWithNotOptimalTransformation,
```
Comment (v-Golubev) [inline inference-engine/tests/functional/plugin/gpu/shared_tests_instances/low_precision_transformations/fake_quantize_with_dq_not_optimal_transformation.cpp:64]: We really need to disable these tests?

File: inference-engine/tests/functional/plugin/gpu/shared_tests_instances/low_precision_transformations/convolution_qdq_transformation.cpp [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+        "U8"
+    },
+};
+
+const std::vector<ngraph::Shape> shapes = {
+    { 1, 3, 16, 16 },
+    { 4, 3, 16, 16 }
+};
+
+INSTANTIATE_TEST_CASE_P(DISABLED_smoke_LPT, ConvolutionQDqTransformation,
```
Comment (v-Golubev) [inline inference-engine/tests/functional/plugin/gpu/shared_tests_instances/low_precision_transformations/convolution_qdq_transformation.cpp:136]: We really need to disable these tests?

File: inference-engine/src/low_precision_transformations/src/reshape.cpp [hunk 9]
```diff
@@ -147,11 +147,15 @@ void reshapeDequantizationConstant(const std::shared_ptr<opset1::Reshape>& resha
 
 bool ReshapeTransformation::transform(TransformationContext& context, ngraph::pattern::Matcher &m) const {
     std::shared_ptr<opset1::Reshape> reshape = as_type_ptr<opset1::Reshape>(m.get_match_root());
-    if ((reshape == nullptr) || (!canBeTransformed(context, reshape))) {
+    if (NetworkHelper::onWeights(reshape)) {
```
Comment (v-Golubev) [inline inference-engine/src/low_precision_transformations/src/reshape.cpp:150]: Please use NetworkHelper::isConstantPath instead of NetworkHelper::onWeights

File: inference-engine/tests/functional/plugin/gpu/shared_tests_instances/low_precision_transformations/fake_quantize_with_dq_not_optimal_transformation.cpp (hunk 7 above)
Comment (eshoguli) [reply inference-engine/tests/functional/plugin/gpu/shared_tests_instances/low_precision_transformations/fake_quantize_with_dq_not_optimal_transformation.cpp]: enabled, thanks

File: inference-engine/tests/functional/plugin/gpu/shared_tests_instances/low_precision_transformations/convolution_qdq_transformation.cpp (hunk 8 above)
Comment (eshoguli) [reply inference-engine/tests/functional/plugin/gpu/shared_tests_instances/low_precision_transformations/convolution_qdq_transformation.cpp]: enabled, thanks

File: inference-engine/src/low_precision_transformations/src/reshape.cpp (hunk 9 above)
Comment (eshoguli) [reply inference-engine/src/low_precision_transformations/src/reshape.cpp]: done, thanks

File: inference-engine/src/transformations/src/transformations/low_precision/disable_convert_constant_folding_on_const_path.cpp [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+            }
+        }
+
+        auto parent = convert->get_input_node_ptr(0);
+        auto child = convert->output(0).get_target_inputs().begin()->get_node();
+        if (is_type<ngraph::opset1::Constant>(parent) &&
+            (is_type<ngraph::opset1::Subtract>(child) || is_type<ngraph::opset1::Multiply>(child))) {
+            auto& rtInfo = convert->get_rt_info();
+            rtInfo["DISABLED_CONSTANT_FOLDING"] = std::make_shared<VariantWrapper<std::string>>("");
+            return true;
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/low_precision/disable_convert_constant_folding_on_const_path.cpp:44]: What the hell is going here?) Please explain

Comment (eshoguli) [reply inference-engine/src/transformations/src/transformations/low_precision/disable_convert_constant_folding_on_const_path.cpp]: comments were added: https://github.com/openvinotoolkit/openvino/pull/3308/files#diff-2caa54ee49daa38e11cd13d5b79b19eb6ae97ab3202f3df6b76099ad9d64635dR49

File: inference-engine/src/transformations/include/transformations/low_precision/disable_convert_constant_folding_on_const_path.hpp (hunk 4 above)
Comment (eshoguli) [reply inference-engine/src/transformations/include/transformations/low_precision/disable_convert_constant_folding_on_const_path.hpp]: moved to LPT, thanks

**Comment by v-Golubev** [comment]
Looks good for me

**Review by vladimir-paramuzov — APPROVED** [review]
gpu changes lgtm

File: inference-engine/src/low_precision_transformations/src/disable_convert_constant_folding_on_const_path.cpp [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+            }
+        }
+
+        // Constant subgraph has not be folded if Convert is part of dequantization operations:
+        //
+        //   Constant                             Constant
+        //      |                                    |
+        //   Convert  Constant           OR       Convert  Constant
+        //       \     /                             \      /
+        //       Subtract   Constant                 Multiply
```
Comment (GlebKazantaev) [inline inference-engine/src/low_precision_transformations/src/disable_convert_constant_folding_on_const_path.cpp:45]: Look like this examples are wrong as you check only Convert child operation. So I suggest to move all this logic to pattern. 
To match multiple types you can use wrap_type:
```cpp
auto eltwise= pattern::wrap_type<opset3::Multiply, opset3::Subtract>(/*inputs*/);
```

File: inference-engine/src/low_precision_transformations/src/disable_convert_constant_folding_on_const_path.cpp [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    auto matcherData = ngraph::pattern::any_input();
+    auto matcherConvert = ngraph::pattern::wrap_type<opset3::Convert>({ matcherData }, pattern::consumers_count(1));
+
+    ngraph::matcher_pass_callback callback = [=](ngraph::pattern::Matcher & m) -> bool {
+        const auto& opsMap = m.get_pattern_value_map();
+        const auto convert = opsMap.find(matcherConvert)->second.get_node()->shared_from_this();
+
+        // validation by Convert operation input precisions
+        if (!inputPrecisions.empty()) {
+            const ngraph::element::Type inputPrecision = convert->input(0).get_element_type();
```
Comment (GlebKazantaev) [inline inference-engine/src/low_precision_transformations/src/disable_convert_constant_folding_on_const_path.cpp:33]: You need to move this check to pattern predicate.

**Review by GlebKazantaev — CHANGES_REQUESTED** [review]

**Review by dmitry-gorokhov — APPROVED** [review]
CPU part looks good to me

**Comment by eshoguli** [comment]
integration branch was merged: https://github.com/openvinotoolkit/openvino/pull/3575, comments were added as tasks
